modulus/metrics/general/wasserstein.py

```diff
@@ -12,12 +12,113 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from warnings import warn
+
 import torch
 
+from modulus.metrics.general import histogram
+
 Tensor = torch.Tensor
 
 
-def wasserstein(bin_edges: Tensor, cdf_x: Tensor, cdf_y: Tensor) -> Tensor:
+def wasserstein_from_normal(
+    mu0: Tensor, sigma0: Tensor, mu1: Tensor, sigma1: Tensor
+) -> Tensor:
+    """Compute the wasserstein distances between two (possibly multivariate) normal
+    distributions.
+
+    Parameters
+    ----------
+    mu0 : Tensor [B (optional), d1]
+        The mean of distribution 0. Can optionally have a batched first dimension.
+    sigma0 : Tensor [B (optional), d1, d2 (optional)]
+        The variance or covariance of distribution 0. If mu0 has a batched dimension,
+        then so must sigma0. If sigma0 is 2 dimension, it is assumed to be a covariance matrix
+        and must be symmetric positive definite.
+    mu1 : Tensor [B (optional), d1]
+        The mean of distribution 1. Can optionally have a batched first dimension.
+    sigma1 : Tensor [B (optional), d1, d2 (optional)]
+        The variance or covariance of distribution 1. If mu1 has a batched dimension,
+        then so must sigma1. If sigma1 is 2 dimension, it is assumed to be a covariance matrix
+        and must be symmetric positive definite.
+
+    Returns
+    -------
+    Tensor [B]
+        The wasserstein distance between N(mu0, sigma0) and N(mu1, sigma1)
+    """
+    mu_ndim = mu0.ndim
+    sigma_ndim = sigma0.ndim
+    if sigma_ndim == mu_ndim:
+        # Univariate normal distribution
+        return (mu0 - mu1) ** 2 + (sigma0 + sigma1 - 2 * torch.sqrt(sigma0 * sigma1))
+
+    else:
+        # Multivariate normal distribution
+        # Compute trace(sig0 + sig1 - 2*(sig0^1/2 * sig1 * sig0^1/2)^1/2) first
+
+        # Compute sig0^1/2 first using eigen decomposition.
+        vals0, vecs0 = torch.linalg.eigh(sigma0)
+        if torch.any(vals0 < 0.0):
+            warn(
+                "Warning! Some eigenvalues are less than zero and matrix is not positive definite."
+            )
+            vals0 = torch.nn.functional.relu(vals0)
+        sqrt_sig0 = torch.matmul(
```

Comment:
Maybe worth considering Newton-Schulz or other series expansion based methods here if performance is a consideration? See this paper for discussion and comparisons to truncated Taylor polynomials and Pade approximants: https://openreview.net/pdf?id=-AOEi-5VTU8

> To give a concrete example, when the batch size is 64, our MPA-Lya is 2.58X faster than NS iteration and 27.25X faster than SVD, while our MTP-Lya is 5.82X faster than the NS iteration and 61.32X faster than SVD.

Reply:
I like this idea, I'll run some benchmarks.

Reply:
Curious how much performance of this calculation is important for the (diffusion?) models that use this. Any ideas?

Reply:
So I implemented Newton-Schulz and ran some tests with random `(100,100)` and `(1000,1000)` matrices. The timings for Newton-Schulz depend on the stopping criterion (some tolerance in successive iterates). The results are compiled from 1000 independent tests and are as follows (in ms):

```
100x100
Eigen Decomposition:
Mean: 3.8977         Std: 4.6846   Low: 2.5036         Max: 148.9305
Newton:
Mean: 2.6360         Std: 16.9361   Low: 1.1909         Max: 536.7705

1000x1000
Eigen Decomposition:
Mean: 25.462         Std: 5.679          Low: 22.706   Max: 201.514
Newton:
Mean: 13.857          Std: 14.799         Low: 12.264         Max: 481.065
```

Reply:
I will include them both in this PR. Maybe this is the start of a new `modulus.linalg`

Reply:
Upon further discussion, this will be postponed to a further PR.